webpack/webpack5-localization-plugin/src/TrueHashPlugin.ts

```diff
@@ -0,0 +1,401 @@
+                }
+              }
+            } else {
+              const assetNames: string[] = Array.from(chunk.files);
+              let jsAssetName: string | undefined;
+              for (const assetName of assetNames) {
+                if (assetName.endsWith('.js')) {
+                  if (jsAssetName) {
+                    compilation.errors.push(
+                      new thisWebpack.WebpackError(`Found multiple .js assets for chunk ${chunk.id}.`)
+                    );
+                  } else {
+                    jsAssetName = assetName;
+                  }
+                }
+              }
+
+              if (!jsAssetName) {
+                compilation.errors.push(
+                  new thisWebpack.WebpackError(`Could not find a .js asset for chunk ${chunk.id}.`)
+                );
+              } else {
+                const { trueHash, newJsFilename } = processChunkAsset(jsAssetName, undefined) ?? {};
+                hashReplacementsByChunk.set(chunk, { existingHash, trueHashByLocale: trueHash });
+                if (newJsFilename) {
+                  chunk.files.delete(jsAssetName);
+                  chunk.files.add(newJsFilename);
+                }
+              }
+            }
+          }
+
+          unprocessedChunks.delete(chunk);
+          const dependents: Set<Chunk> | undefined = dependentsByChunk.get(chunk);
+          if (dependents) {
+            for (const dependent of dependents) {
+              unprocessedDependenciesByChunk.get(dependent)?.delete(chunk);
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
+/**
+ * @public
+ */
+export class TrueHashPlugin implements WebpackPluginInstance {
+  private readonly _options: ITrueHashPluginOptions;
+
+  public constructor(options: ITrueHashPluginOptions = {}) {
+    this._options = options;
+  }
+
+  public apply(compiler: Compiler): void {
+    compiler.hooks.thisCompilation.tap(PLUGIN_NAME, (compilation: Compilation) => {
+      const { webpack: thisWebpack } = compiler;
+
+      let hasLocalizationPluginTrueHashOption: boolean = false;
+      if (compiler.options.plugins) {
+        for (const plugin of compiler.options.plugins) {
+          if (plugin instanceof LocalizationPlugin && plugin._options.useTrueHashes) {
+            hasLocalizationPluginTrueHashOption = true;
+            break;
+          }
+        }
+      }
+
+      if (hasLocalizationPluginTrueHashOption) {
+        compilation.warnings.push(
+          new thisWebpack.WebpackError(
+            `The ${TrueHashPlugin.name} is not compatible with the LocalizationPlugin's "useTrueHashes" option. ` +
+              `Because the LocalizationPlugin is already handling true hashes, the ${TrueHashPlugin.name} plugin ` +
+              'will have no effect.'
+          )
+        );
+      } else {
+        const { stageOverride: processAssetsStage = thisWebpack.Compilation.PROCESS_ASSETS_STAGE_SUMMARIZE } =
```

Comment:
I think this should default to some value < STAGE_OPTIMIZE_TRANSFER

Reply:
Changing this to `STAGE_OPTIMIZE_TRANSFER - 1` doesn't work. It can't find the assets for chunks at that point.

Reply:
This isn't being published anyway, so it doesn't matter.